### PR TITLE
fix(router): associate parent payment token with `payment_method_id` as hyperswitch token for saved cards

### PR DESCRIPTION
**Pull request by vspecky**
## Type of Change

- [x] Bugfix
- [ ] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

This PR implements some missing components of #2073 , mainly the logic to associate the generated parent payment token with the `payment_method_id` as opposed to a randomly generated temp locker token in the saved cards flow, so a call can be made directly to the persistent locker to fetch card details.

### Additional Changes
None
- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context

Due to the missing association between parent payment token and the payment method id as hyperswitch token for saved cards, the saved cards flow had some potential issues which needed to be addressed.

## How did you test it?

Local testing of happy flows and edge cases, compiler-guided.

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/types/storage/payment_method.rs [hunk 1]
```diff
@@ -2,3 +2,32 @@ pub use diesel_models::payment_method::{
     PaymentMethod, PaymentMethodNew, PaymentMethodUpdate, PaymentMethodUpdateInternal,
     TokenizeCoreWorkflow,
 };
+
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+#[serde(rename_all = "snake_case")]
+pub enum HypsTokenKind {
```
Comment (jarnura) [inline crates/router/src/types/storage/payment_method.rs:8]: Avoid using hyperswitch and hyps in codebase

Comment (jarnura) [inline crates/router/src/types/storage/payment_method.rs:8]:
```suggestion
pub enum RouterTokenKind {
```

File: crates/router/src/types/storage/payment_method.rs [hunk 2]
```diff
@@ -2,3 +2,32 @@ pub use diesel_models::payment_method::{
...
+    Temporary,
+    Permanent,
+}
+
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+pub struct HyperswitchTokenData {
```
Comment (jarnura) [inline crates/router/src/types/storage/payment_method.rs:14]:
```suggestion
pub struct RouterTokenData {
```

File: crates/router/src/core/payments/helpers.rs [hunk 3]
```diff
@@ -1227,72 +1246,122 @@ pub async fn make_pm_data<'a, F: Clone, R>(
...
-        (_, Some(hyperswitch_token)) => {
-            let (pm, supplementary_data) = vault::Vault::get_payment_method_data_from_locker(
-                state,
-                &hyperswitch_token,
-            )
-            .await
-            .attach_printable(
-                "Payment method for given token not found or there was a problem fetching it",
-            )?;
+        (_, Some(hyperswitch_token)) => match hyperswitch_token.kind {
```
Comment (jarnura) [inline crates/router/src/core/payments/helpers.rs:1249]: Avoid writing long implementation, split branching into separate functions and writing comments for those.

**Review by jarnura — CHANGES_REQUESTED** [review]
Don't

**Review by jarnura — DISMISSED** [review]

**Review by Sarthak1799 — APPROVED** [review]

File: crates/router/src/core/payments/helpers.rs [hunk 4]
```diff
@@ -1345,7 +1454,26 @@ pub async fn make_pm_data<'a, F: Clone, R, Ctx: PaymentMethodRetrieve>(
                         },
                     ))?;
 
-                Some(key)
+                let token_data_result = token_data_string
+                    .clone()
+                    .parse_struct("HyperswitchTokenData")
+                    .change_context(errors::ApiErrorResponse::InternalServerError)
+                    .attach_printable("failed to deserialize hyperswitch token data");
```
Comment (Sarthak1799) [inline crates/router/src/core/payments/helpers.rs:1461]: So token is being serialized as a `PaymentTokenData` and expected to be parsed as `HyperswitchTokenData`?
So is the other struct for old entries in redis?

**Review by Sarthak1799 — APPROVED** [review]

**Review by jarnura — APPROVED** [review]

**Comment by vspecky** [comment]
Test Case :-
1. Saved cards flow should succeed
2. Temporary tokenization flow should succeed (card details given during Payments Create, then Payment Token given during Confirm)

**Comment by Sarthak1799** [comment]
Could you resolve the conflicts on the linked PR?

**Review by jarnura — APPROVED** [review]

**Review by Sarthak1799 — APPROVED** [review]

**Pull request merged**